File: LicenseManager.m
//
//  LicenseManager.m
//  Sistema de Licenciamento para iOS
//
//  Gerencia validação de licenças contra API remota
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import <Security/Security.h>
#import <CommonCrypto/CommonDigest.h>
#import <sys/sysctl.h>

static NSString * const kKeychainService = @"com.martinsmod.license";
static NSString * const kKeychainAccount = @"license_key";

// MARK: - Estrutura de Dados

typedef void (^LicenseCompletionHandler)(BOOL success, NSString * _Nullable error);
typedef void (^LicenseCheckHandler)(BOOL valid, NSString * _Nullable error);

// MARK: - Classe LicenseManager

@interface LicenseManager : NSObject
@property (nonatomic, strong) NSString *apiBaseURL;
@property (nonatomic, strong) NSTimer *verificationTimer;
@end

@implementation LicenseManager

static LicenseManager *sharedInstance = nil;

+ (instancetype)sharedManager {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[LicenseManager alloc] initWithAPIURL:@"https://187.127.45.32:3000"];
    });
    return sharedInstance;
}

- (instancetype)initWithAPIURL:(NSString *)apiURL {
    self = [super init];
    if (self) {
        self.apiBaseURL = apiURL;
        [self setupVerificationTimer];
    }
    return self;
}

// MARK: - Ativar Licença

- (void)activateLicense:(NSString *)key completion:(LicenseCompletionHandler)completion {
    if (![self isValidKeyFormat:key]) {
        if (completion) {
            completion(NO, @"Formato de chave inválido");
        }
        return;
    }
    
    NSString *deviceId = [self getDeviceIdentifier];
    NSString *appVersion = [self getAppVersion];
    NSString *osVersion = [[UIDevice currentDevice] systemVersion];
    
    NSDictionary *requestBody = @{
        @"key": key,
        @"deviceId": deviceId,
        @"appVersion": appVersion,
        @"osVersion": osVersion
    };
    
    [self makeAPIRequestWithEndpoint:@"/api/license/activate"
                              method:@"POST"
                                body:requestBody
                          completion:^(NSDictionary *response, NSError *error) {
        if (error) {
            if (completion) {
                completion(NO, error.localizedDescription);
            }
            return;
        }
        
        NSNumber *valid = response[@"valid"];
        if (valid && [valid boolValue]) {
            [self storeKeyInKeychain:key];
            if (completion) {
                completion(YES, nil);
            }
        } else {
            NSString *errorCode = response[@"error"] ?: @"Erro desconhecido";
            if (completion) {
                completion(NO, errorCode);
            }
        }
    }];
}

// MARK: - Verificar Licença

- (void)checkLicenseWithCompletion:(LicenseCheckHandler)completion {
    NSString *key = [self retrieveKeyFromKeychain];
    if (!key) {
        if (completion) {
            completion(NO, @"LICENSE_NOT_FOUND");
        }
        return;
    }
    
    NSString *deviceId = [self getDeviceIdentifier];
    
    NSDictionary *requestBody = @{
        @"key": key,
        @"deviceId": deviceId
    };
    
    [self makeAPIRequestWithEndpoint:@"/api/license/check"
                              method:@"POST"
                                body:requestBody
                          completion:^(NSDictionary *response, NSError *error) {
        if (error) {
            if (completion) {
                completion(NO, error.localizedDescription);
            }
            return;
        }
        
        NSNumber *valid = response[@"valid"];
        if (valid && [valid boolValue]) {
            if (completion) {
                completion(YES, nil);
            }
        } else {
            NSString *errorCode = response[@"error"] ?: @"Erro desconhecido";
            if (completion) {
                completion(NO, errorCode);
            }
        }
    }];
}

// MARK: - Obter Informações

- (void)getRemainingSecondsWithCompletion:(void (^)(NSInteger seconds, NSString * _Nullable error))completion {
    NSString *key = [self retrieveKeyFromKeychain];
    if (!key) {
        if (completion) {
            completion(0, @"LICENSE_NOT_FOUND");
        }
        return;
    }
    
    NSString *deviceId = [self getDeviceIdentifier];
    
    NSDictionary *requestBody = @{
        @"key": key,
        @"deviceId": deviceId
    };
    
    [self makeAPIRequestWithEndpoint:@"/api/license/check"
                              method:@"POST"
                                body:requestBody
                          completion:^(NSDictionary *response, NSError *error) {
        if (error) {
            if (completion) {
                completion(0, error.localizedDescription);
            }
            return;
        }
        
        NSNumber *remainingSeconds = response[@"remainingSeconds"];
        if (remainingSeconds) {
            if (completion) {
                completion([remainingSeconds integerValue], nil);
            }
        } else {
            if (completion) {
                completion(0, @"Erro ao obter tempo restante");
            }
        }
    }];
}

- (void)getExpirationDateWithCompletion:(void (^)(NSDate * _Nullable date, NSString * _Nullable error))completion {
    NSString *key = [self retrieveKeyFromKeychain];
    if (!key) {
        if (completion) {
            completion(nil, @"LICENSE_NOT_FOUND");
        }
        return;
    }
    
    NSString *deviceId = [self getDeviceIdentifier];
    
    NSDictionary *requestBody = @{
        @"key": key,
        @"deviceId": deviceId
    };
    
    [self makeAPIRequestWithEndpoint:@"/api/license/check"
                              method:@"POST"
                                body:requestBody
                          completion:^(NSDictionary *response, NSError *error) {
        if (error) {
            if (completion) {
                completion(nil, error.localizedDescription);
            }
            return;
        }
        
        NSString *expiresAtString = response[@"expiresAt"];
        if (expiresAtString) {
            NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
            formatter.dateFormat = @"yyyy-MM-dd'T'HH:mm:ss'Z'";
            formatter.timeZone = [NSTimeZone timeZoneWithName:@"UTC"];
            NSDate *expiresAt = [formatter dateFromString:expiresAtString];
            
            if (completion) {
                completion(expiresAt, nil);
            }
        } else {
            if (completion) {
                completion(nil, @"Erro ao obter data de expiração");
            }
        }
    }];
}

// MARK: - Desativar Licença

- (void)deactivateLicense {
    [self removeKeyFromKeychain];
    [self stopVerificationTimer];
}

// MARK: - Propriedades

- (BOOL)isLicenseValid {
    NSString *key = [self retrieveKeyFromKeychain];
    return key != nil && [self isValidKeyFormat:key];
}

// MARK: - Verificação Periódica

- (void)setupVerificationTimer {
    NSTimeInterval interval = 6 * 60 * 60; // 6 horas
    self.verificationTimer = [NSTimer scheduledTimerWithTimeInterval:interval
                                                              target:self
                                                            selector:@selector(verifyLicensePeriodically)
                                                            userInfo:nil
                                                             repeats:YES];
}

- (void)stopVerificationTimer {
    [self.verificationTimer invalidate];
    self.verificationTimer = nil;
}

- (void)verifyLicensePeriodically {
    [self checkLicenseWithCompletion:^(BOOL valid, NSString *error) {
        if (!valid) {
            NSLog(@"🚨 LICENÇA INVÁLIDA: %@", error);
            [self crashApp:error ?: @"Licença inválida"];
        }
    }];
}

// MARK: - Métodos Privados

- (BOOL)isValidKeyFormat:(NSString *)key {
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"^[A-Z0-9]{4}-[A-Z0-9]{4}-[A-Z0-9]{4}-[A-Z0-9]{4}$"
                                                                           options:0
                                                                             error:nil];
    NSRange range = NSMakeRange(0, key.length);
    return [regex numberOfMatchesInString:key options:0 range:range] > 0;
}

- (NSString *)getDeviceIdentifier {
    char buffer[256];
    size_t bufferSize = sizeof(buffer);
    int result = sysctlbyname("hw.machine", buffer, &bufferSize, NULL, 0);
    
    NSString *machine = @"unknown";
    if (result == 0) {
        machine = [NSString stringWithCString:buffer encoding:NSUTF8StringEncoding];
    }
    
    NSString *uuid = [[UIDevice currentDevice].identifierForVendor UUIDString] ?: @"unknown";
    NSString *combined = [NSString stringWithFormat:@"%@-%@", machine, uuid];
    
    return [self hashSHA256:combined];
}

- (NSString *)getAppVersion {
    NSString *version = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"CFBundleShortVersionString"];
    return version ?: @"1.0.0";
}

- (NSString *)hashSHA256:(NSString *)input {
    NSData *data = [input dataUsingEncoding:NSUTF8StringEncoding];
    unsigned char digest[CC_SHA256_DIGEST_LENGTH];
    CC_SHA256(data.bytes, (CC_LONG)data.length, digest);
    
    NSMutableString *result = [NSMutableString stringWithCapacity:CC_SHA256_DIGEST_LENGTH * 2];
    for (int i = 0; i < CC_SHA256_DIGEST_LENGTH; i++) {
        [result appendFormat:@"%02x", digest[i]];
    }
    
    return result;
}

// MARK: - Keychain

- (void)storeKeyInKeychain:(NSString *)key {
    NSData *keyData = [key dataUsingEncoding:NSUTF8StringEncoding];
    
    NSDictionary *query = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrService: kKeychainService,
        (__bridge id)kSecAttrAccount: kKeychainAccount,
        (__bridge id)kSecValueData: keyData
    };
    
    SecItemDelete((__bridge CFDictionaryRef)query);
    SecItemAdd((__bridge CFDictionaryRef)query, NULL);
}

- (NSString *)retrieveKeyFromKeychain {
    NSDictionary *query = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrService: kKeychainService,
        (__bridge id)kSecAttrAccount: kKeychainAccount,
        (__bridge id)kSecReturnData: @YES
    };
    
    CFDataRef result = NULL;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, (CFTypeRef *)&result);
    
    if (status == errSecSuccess && result) {
        NSString *key = [[NSString alloc] initWithData:(__bridge NSData *)result encoding:NSUTF8StringEncoding];
        CFRelease(result);
        return key;
    }
    
    return nil;
}

- (void)removeKeyFromKeychain {
    NSDictionary *query = @{
        (__bridge id)kSecClass: (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrService: kKeychainService,
        (__bridge id)kSecAttrAccount: kKeychainAccount
    };
    
    SecItemDelete((__bridge CFDictionaryRef)query);
}

// MARK: - API

- (void)makeAPIRequestWithEndpoint:(NSString *)endpoint
                            method:(NSString *)method
                              body:(NSDictionary *)body
                        completion:(void (^)(NSDictionary *response, NSError *error))completion {
    NSString *urlString = [NSString stringWithFormat:@"%@%@", self.apiBaseURL, endpoint];
    NSURL *url = [NSURL URLWithString:urlString];
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = method;
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    
    if (body) {
        NSError *jsonError = nil;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:body options:0 error:&jsonError];
        if (!jsonError) {
            request.HTTPBody = jsonData;
        }
    }
    
    NSURLSessionDataTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request
                                                                   completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error) {
            if (completion) {
                completion(nil, error);
            }
            return;
        }
        
        NSError *jsonError = nil;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
        
        if (completion) {
            completion(json, jsonError);
        }
    }];
    
    [task resume];
}

// MARK: - Crash

- (void)crashApp:(NSString *)reason {
    NSLog(@"🚨 LICENÇA INVÁLIDA: %@", reason);
    
    NSException *exception = [NSException exceptionWithName:@"LicenseInvalidException"
                                                     reason:reason
                                                   userInfo:nil];
    @throw exception;
}

@end

File: dylib/Sources/LicenseManager.swift
import Foundation
import Security

/**
 LicenseManager - Gerenciador de Licenças para iOS
 Responsável por validar, armazenar e verificar licenças de forma segura
 */
public class LicenseManager {
    static let shared = LicenseManager()
    
    private let keychainService = "com.martinsmod.license"
    private let keychainAccount = "license_key"
    private let apiBaseURL: String
    private var verificationTimer: Timer?
    private let verificationInterval: TimeInterval = 6 * 60 * 60 // 6 horas
    
    public var isLicenseValid: Bool {
        guard let key = retrieveKeyFromKeychain() else { return false }
        return validateLicenseLocally(key)
    }
    
    public init(apiURL: String = "https://187.127.45.32:3000") {
        self.apiBaseURL = apiURL
        setupVerificationTimer()
    }
    
    // MARK: - Public Methods
    
    /**
     Ativa uma licença no dispositivo
     */
    public func activateLicense(_ key: String, completion: @escaping (Bool, String?) -> Void) {
        guard isValidKeyFormat(key) else {
            completion(false, "Formato de chave inválido")
            return
        }
        
        let deviceId = getDeviceIdentifier()
        let appVersion = getAppVersion()
        let osVersion = UIDevice.current.systemVersion
        
        let requestBody: [String: Any] = [
            "key": key,
            "deviceId": deviceId,
            "appVersion": appVersion,
            "osVersion": osVersion
        ]
        
        makeAPIRequest(
            endpoint: "/api/license/activate",
            method: "POST",
            body: requestBody
        ) { [weak self] response, error in
            guard let self = self else { return }
            
            if let error = error {
                completion(false, error)
                return
            }
            
            guard let response = response as? [String: Any],
                  let valid = response["valid"] as? Bool else {
                completion(false, "Resposta inválida da API")
                return
            }
            
            if valid {
                // Armazena a chave no Keychain
                self.storeKeyInKeychain(key)
                completion(true, nil)
            } else {
                let errorCode = response["error"] as? String ?? "Erro desconhecido"
                completion(false, errorCode)
            }
        }
    }
    
    /**
     Verifica se a licença é válida
     */
    public func checkLicense(completion: @escaping (Bool, String?) -> Void) {
        guard let key = retrieveKeyFromKeychain() else {
            completion(false, "LICENSE_NOT_FOUND")
            return
        }
        
        let deviceId = getDeviceIdentifier()
        
        let requestBody: [String: Any] = [
            "key": key,
            "deviceId": deviceId
        ]
        
        makeAPIRequest(
            endpoint: "/api/license/check",
            method: "POST",
            body: requestBody
        ) { response, error in
            if let error = error {
                completion(false, error)
                return
            }
            
            guard let response = response as? [String: Any],
                  let valid = response["valid"] as? Bool else {
                completion(false, "Resposta inválida da API")
                return
            }
            
            if valid {
                completion(true, nil)
            } else {
                let errorCode = response["error"] as? String ?? "Erro desconhecido"
                completion(false, errorCode)
            }
        }
    }
    
    /**
     Obtém o tempo restante da licença em segundos
     */
    public func getRemainingSeconds(completion: @escaping (Int?, String?) -> Void) {
        guard let key = retrieveKeyFromKeychain() else {
            completion(nil, "LICENSE_NOT_FOUND")
            return
        }
        
        let deviceId = getDeviceIdentifier()
        
        let requestBody: [String: Any] = [
            "key": key,
            "deviceId": deviceId
        ]
        
        makeAPIRequest(
            endpoint: "/api/license/check",
            method: "POST",
            body: requestBody
        ) { response, error in
            if let error = error {
                completion(nil, error)
                return
            }
            
            guard let response = response as? [String: Any],
                  let remainingSeconds = response["remainingSeconds"] as? Int else {
                completion(nil, "Erro ao obter tempo restante")
                return
            }
            
            completion(remainingSeconds, nil)
        }
    }
    
    /**
     Obtém a data de expiração da licença
     */
    public func getExpirationDate(completion: @escaping (Date?, String?) -> Void) {
        guard let key = retrieveKeyFromKeychain() else {
            completion(nil, "LICENSE_NOT_FOUND")
            return
        }
        
        let deviceId = getDeviceIdentifier()
        
        let requestBody: [String: Any] = [
            "key": key,
            "deviceId": deviceId
        ]
        
        makeAPIRequest(
            endpoint: "/api/license/check",
            method: "POST",
            body: requestBody
        ) { response, error in
            if let error = error {
                completion(nil, error)
                return
            }
            
            guard let response = response as? [String: Any],
                  let expiresAtString = response["expiresAt"] as? String else {
                completion(nil, "Erro ao obter data de expiração")
                return
            }
            
            let formatter = ISO8601DateFormatter()
            if let expiresAt = formatter.date(from: expiresAtString) {
                completion(expiresAt, nil)
            } else {
                completion(nil, "Formato de data inválido")
            }
        }
    }
    
    /**
     Desativa a licença (remove do Keychain)
     */
    public func deactivateLicense() {
        removeKeyFromKeychain()
        stopVerificationTimer()
    }
    
    // MARK: - Private Methods
    
    private func setupVerificationTimer() {
        verificationTimer = Timer.scheduledTimer(withTimeInterval: verificationInterval, repeats: true) { [weak self] _ in
            self?.verifyLicensePeriodically()
        }
    }
    
    private func stopVerificationTimer() {
        verificationTimer?.invalidate()
        verificationTimer = nil
    }
    
    private func verifyLicensePeriodically() {
        checkLicense { [weak self] valid, error in
            if !valid {
                // Licença inválida, crashea o app
                self?.crashApp(reason: error ?? "Licença inválida")
            }
        }
    }
    
    private func validateLicenseLocally(_ key: String) -> Bool {
        return isValidKeyFormat(key)
    }
    
    private func isValidKeyFormat(_ key: String) -> Bool {
        let keyRegex = "^[A-Z0-9]{4}-[A-Z0-9]{4}-[A-Z0-9]{4}-[A-Z0-9]{4}$"
        let predicate = NSPredicate(format: "SELF MATCHES %@", keyRegex)
        return predicate.evaluate(with: key)
    }
    
    private func getDeviceIdentifier() -> String {
        // Gera um hash do identificador único do dispositivo
        var systemInfo = utsname()
        uname(&systemInfo)
        
        let machineMirror = Mirror(reflecting: systemInfo.machine)
        let identifier = machineMirror.children.reduce("") { identifier, element in
            guard let value = element.value as? Int8, value != 0 else { return identifier }
            return identifier + String(UnicodeScalar(UInt8(value)))
        }
        
        // Combina com UUID do dispositivo
        let uuid = UIDevice.current.identifierForVendor?.uuidString ?? "unknown"
        let combined = "\(identifier)-\(uuid)"
        
        return hashSHA256(combined)
    }
    
    private func getAppVersion() -> String {
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0"
        return version
    }
    
    private func hashSHA256(_ input: String) -> String {
        let data = input.data(using: .utf8) ?? Data()
        var digest = [UInt8](repeating: 0, count: Int(32))
        
        #if os(iOS)
        import CommonCrypto
        data.withUnsafeBytes { buffer in
            _ = CC_SHA256(buffer.baseAddress, CC_LONG(data.count), &digest)
        }
        #else
        // Fallback para simulador
        digest = Array(data.prefix(32))
        #endif
        
        return digest.map { String(format: "%02x", $0) }.joined()
    }
    
    // MARK: - Keychain Methods
    
    private func storeKeyInKeychain(_ key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount,
            kSecValueData as String: key.data(using: .utf8) ?? Data()
        ]
        
        SecItemDelete(query as CFDictionary)
        SecItemAdd(query as CFDictionary, nil)
    }
    
    private func retrieveKeyFromKeychain() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount,
            kSecReturnData as String: true
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        guard status == errSecSuccess,
              let data = result as? Data,
              let key = String(data: data, encoding: .utf8) else {
            return nil
        }
        
        return key
    }
    
    private func removeKeyFromKeychain() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: keychainService,
            kSecAttrAccount as String: keychainAccount
        ]
        
        SecItemDelete(query as CFDictionary)
    }
    
    // MARK: - API Methods
    
    private func makeAPIRequest(
        endpoint: String,
        method: String = "GET",
        body: [String: Any]? = nil,
        completion: @escaping ([String: Any]?, String?) -> Void
    ) {
        guard let url = URL(string: apiBaseURL + endpoint) else {
            completion(nil, "URL inválida")
            return
        }
        
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        if let body = body {
            do {
                request.httpBody = try JSONSerialization.data(withJSONObject: body)
            } catch {
                completion(nil, "Erro ao serializar JSON")
                return
            }
        }
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(nil, error.localizedDescription)
                return
            }
            
            guard let data = data else {
                completion(nil, "Sem dados na resposta")
                return
            }
            
            do {
                if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                    completion(json, nil)
                } else {
                    completion(nil, "Resposta não é um JSON válido")
                }
            } catch {
                completion(nil, "Erro ao decodificar JSON")
            }
        }
        
        task.resume()
    }
    
    // MARK: - Crash Methods
    
    private func crashApp(reason: String) {
        print("🚨 LICENÇA INVÁLIDA: \(reason)")
        
        // Força o crash do app
        let exception = NSException(name: NSExceptionName("LicenseInvalidException"), reason: reason, userInfo: nil)
        exception.raise()
    }
}

// MARK: - UIDevice Extension

extension UIDevice {
    var identifierForVendor: UUID? {
        return UIDevice.current.identifierForVendor
    }
}
